Description: Hide window buttons when running in Unity and the window is maximized.
 Plus define the "toolbar-mode" style class on maximized windows.
Author: Marco Trevisan <marco.trevisan@canonical.com>
Bug-Ubuntu: https://launchpad.net/bugs/1515810

Index: gtk+3.0-3.18.8/gtk/gtkheaderbar.c
===================================================================
--- gtk+3.0-3.18.8.orig/gtk/gtkheaderbar.c
+++ gtk+3.0-3.18.8/gtk/gtkheaderbar.c
@@ -31,6 +31,10 @@
 
 #include <string.h>
 
+#ifdef GDK_WINDOWING_X11
+#include "x11/gdkx.h"
+#endif
+
 /**
  * SECTION:gtkheaderbar
  * @Short_description: A box with a centered child
@@ -65,6 +69,7 @@ struct _GtkHeaderBarPrivate
   GtkWidget *custom_title;
   gint spacing;
   gboolean has_subtitle;
+  gboolean unity_environment;
 
   GList *children;
 
@@ -311,6 +316,21 @@ _gtk_header_bar_update_window_buttons (G
   if (!priv->shows_wm_decorations)
     return;
 
+  window = GTK_WINDOW (gtk_widget_get_toplevel (widget));
+
+  if (priv->unity_environment)
+    {
+      if (gtk_window_is_maximized (window))
+        {
+          gtk_widget_hide (priv->title_label);
+          gtk_style_context_add_class (gtk_widget_get_style_context (widget), "toolbar-mode");
+          return;
+        }
+
+      gtk_widget_show (priv->title_label);
+      gtk_style_context_remove_class (gtk_widget_get_style_context (widget), "toolbar-mode");
+    }
+
   direction = gtk_widget_get_direction (widget);
 
   g_object_get (gtk_widget_get_settings (widget),
@@ -324,7 +344,6 @@ _gtk_header_bar_update_window_buttons (G
       layout_desc = g_strdup (priv->decoration_layout);
     }
 
-  window = GTK_WINDOW (gtk_widget_get_toplevel (widget));
 
   if (!shown_by_shell && gtk_window_get_application (window))
     menu = gtk_application_get_app_menu (gtk_window_get_application (window));
@@ -1808,6 +1827,22 @@ gtk_header_bar_realize (GtkWidget *widge
 
   GTK_WIDGET_CLASS (gtk_header_bar_parent_class)->realize (widget);
 
+#ifdef GDK_WINDOWING_X11
+  GtkHeaderBar *bar;
+  GtkHeaderBarPrivate *priv;
+  GdkScreen *screen;
+
+  bar = GTK_HEADER_BAR (widget);
+  priv = gtk_header_bar_get_instance_private (bar);
+  screen = gtk_widget_get_screen (widget);
+
+  if (GDK_IS_X11_SCREEN (screen))
+    {
+      GdkAtom unity_atom = gdk_atom_intern_static_string ("_UNITY_SHELL");
+      priv->unity_environment = gdk_x11_screen_supports_net_wm_hint (screen, unity_atom);
+    }
+#endif
+
   settings = gtk_widget_get_settings (widget);
   g_signal_connect_swapped (settings, "notify::gtk-shell-shows-app-menu",
                             G_CALLBACK (_gtk_header_bar_update_window_buttons), widget);
